Extension/src/LanguageServer/cppBuildTaskProvider.ts

```diff
@@ -0,0 +1,346 @@
+            ), isCl ? '$msCompile' : '$gcc');
+
+        task.group = vscode.TaskGroup.Build;
+
+        return task;
+    };
+
+    async ensureBuildTaskExists(taskName: string): Promise<void> {
+        const rawTasksJson: any = await getRawTasksJson();
+
+        // Ensure that the task exists in the user's task.json. Task will not be found otherwise.
+        if (!rawTasksJson.tasks) {
+            rawTasksJson.tasks = new Array();
+        }
+        // Find or create the task which should be created based on the selected "debug configuration".
+        let selectedTask: vscode.Task | undefined = rawTasksJson.tasks.find((task: any) => task.label && task.label === taskName);
+        if (selectedTask) {
+            return;
+        }
+
+        const buildTasks: vscode.Task[] = await this.getTasks(true);
+        selectedTask = buildTasks.find(task => task.name === taskName);
+        console.assert(selectedTask);
+        if (!selectedTask) {
+            throw new Error("Failed to get selectedTask in ensureBuildTaskExists()");
+        }
+
+        rawTasksJson.version = "2.0.0";
+
+        const selectedTask2: vscode.Task = selectedTask;
+        if (!rawTasksJson.tasks.find((task: any) => task.label === selectedTask2.definition.label)) {
+            const task: any = {
+                ...selectedTask2.definition,
+                problemMatcher: selectedTask2.problemMatchers,
+                group: { kind: "build", "isDefault": true }
+            };
+            rawTasksJson.tasks.push(task);
+        }
+
+        // TODO: It's dangerous to overwrite this file. We could be wiping out comments.
+        const settings: OtherSettings = new OtherSettings();
+        const tasksJsonPath: string | undefined = getTasksJsonPath();
+        if (!tasksJsonPath) {
+            throw new Error("Failed to get tasksJsonPath in ensureBuildTaskExists()");
+        }
+
+        await util.writeFileText(tasksJsonPath, JSON.stringify(rawTasksJson, null, settings.editorTabSize));
+    }
+}
+
+class CustomBuildTaskTerminal implements vscode.Pseudoterminal {
+    private writeEmitter  = new vscode.EventEmitter<string>();
+    private closeEmitter = new vscode.EventEmitter<number>();
+    public get onDidWrite(): vscode.Event<string> { return this.writeEmitter.event; }
+    public get onDidClose(): vscode.Event<number> { return this.closeEmitter.event; }
+
+    constructor(private command: string, private args: string[], private options: cp.ExecOptions | undefined) {
+    }
+
+
+    open(_initialDimensions: vscode.TerminalDimensions | undefined): void {
+        telemetry.logLanguageServerEvent("cppBuildTaskStarted");
+        // At this point we can start using the terminal.
+        this.writeEmitter.fire("Starting build...\r\n");
+        this.doBuild();
+    }
+
+    close(): void {
+        // The terminal has been closed. Shutdown the build.
+    }
+
+    private async doBuild(): Promise<number> {
```

Comment:
Since this function does not 'await' any other functions, I don't believe it needs to be declared as `async`.  There is also a `execChildProcess` in common.ts which converts `exec` into an await-able function.  You could either a) remove the `await`, or b) await `execChildProcess` instead of invoking `exec` directly.  B seems better, as you could avoid creating a new Promise object directly, and use `try`/`catch`.

Also, where `doBuild` is called, the Promise you are returning is not awaited.  Is the intention to complete the `exec` before `open` returns?  If so, you'll want to await the result of doBuild.  Or, if not, doBuild() doesn't need to return a promise at all or be async.

Reply:
I added an await on the `exec` directly. It was a good idea to re-use `execChildProcess` instead, but it had a `vscode.OutputChannel` as input, so I couldn't re-use it. 
I added `await` to the result of `dobuild` as well.

Reply:
The question mark in `channel?: vscode.OutputChannel` indicates the channel param is optional.

Reply:
I need to print the stdout (instead of error message) in the case that exec finishes with error.

Reply:
It could be updated to return that also.  :)  Generally, I'm suggesting that creation of Promises are pushed down into helper functions, as async/await syntax is so much easier to read and understand.  Not critical, though.